src/coreclr/src/vm/jitinterface.cpp

```diff
@@ -3845,13 +3812,6 @@ BOOL CEEInfo::canInlineTypeCheckWithObjectVTable (CORINFO_CLASS_HANDLE clsHnd)
     _ASSERTE(clsHnd);
 
     TypeHandle VMClsHnd(clsHnd);
-
-    if (VMClsHnd.IsTypeDesc())
-    {
-        // We can't do this optimization for arrays because of the object methodtable is template methodtable
-        ret = FALSE;
-    }
-    else
     if (VMClsHnd == TypeHandle(g_pCanonMethodTableClass))
```

Comment:
Logically this function should be able to simply return TRUE in all cases, but I'm not sure if the jit can tolerate that in the case of a shared generic.. If it can't, please file an issue against the jit about that.

(The interesting case is)

if (typeof(T) == obj.GetType()) {...}

Reply:
`typeof(T) == obj.GetType()`  does not get here. 
`gtFoldTypeCompare` requires that at least one comparand is effectively a constant (`TPK_Handle`) and `typeof(T)` for a generic typeparam is not a constant.

In addition to that, it is actually classified as `TPK_Unknown`. No idea why, I did not dig further.
I would expect `TPK_GetType`.

I agree this is now optimizable. - as long as shared generics are used only for reference types.  (since arrays are no longer special).

I will log a bug for the JIT.

Reply:
Logged [https://github.com/dotnet/runtime/issues/1258](https://github.com/dotnet/runtime/issues/1258)


---
In reply to: [362655670](https://github.com/dotnet/runtime/pull/1201#discussion_r362655670) [](ancestors = 362655670)